src/features.c

```diff
@@ -0,0 +1,344 @@
+  add_array_to_json (json_gen, "actions", linux->seccomp.actions);
+  add_array_to_json (json_gen, "operators", linux->seccomp.operators);
+  add_array_to_json (json_gen, "archs", linux->seccomp.archs);
+
+  yajl_gen_map_close (json_gen);
+}
+
+void
+crun_features_add_apparmor_info (yajl_gen json_gen, const struct linux_info_s *linux)
+{
+  yajl_gen_string (json_gen, (const unsigned char *) "apparmor", strlen ("apparmor"));
+  yajl_gen_map_open (json_gen);
+
+  add_bool_to_json (json_gen, "enabled", linux->apparmor.enabled);
+
+  // Close the map for apparmor
+  yajl_gen_map_close (json_gen);
+}
+
+void
+crun_features_add_selinux_info (yajl_gen json_gen, const struct linux_info_s *linux)
+{
+  yajl_gen_string (json_gen, (const unsigned char *) "selinux", strlen ("selinux"));
+  yajl_gen_map_open (json_gen);
+
+  add_bool_to_json (json_gen, "enabled", linux->selinux.enabled);
+
+  // Close the map for selinux
+  yajl_gen_map_close (json_gen);
+}
+
+void
+crun_features_add_linux_info (yajl_gen json_gen, const struct linux_info_s *linux)
+{
+  yajl_gen_string (json_gen, (const unsigned char *) "linux", strlen ("linux"));
+  yajl_gen_map_open (json_gen);
+
+  crun_features_add_namespaces (json_gen, linux);
+  crun_features_add_capabilities (json_gen, linux);
+  crun_features_add_cgroup_info (json_gen, linux);
+  crun_features_add_seccomp_info (json_gen, linux);
+  crun_features_add_apparmor_info (json_gen, linux);
+  crun_features_add_selinux_info (json_gen, linux);
+
+  yajl_gen_map_close (json_gen);
+}
+
+void
+crun_features_add_annotations_info (yajl_gen json_gen, const struct annotations_info_s *annotation)
+{
+  yajl_gen_string (json_gen, (const unsigned char *) "annotations", strlen ("annotations"));
+  yajl_gen_map_open (json_gen);
+
+  add_string_to_json (json_gen, "io.github.seccomp.libseccomp.version", annotation->io_github_seccomp_libseccomp_version);
+  add_bool_to_json (json_gen, "org.opencontainers.runc.checkpoint.enabled", annotation->org_opencontainers_runc_checkpoint_enabled);
+  add_string_to_json (json_gen, "io.github.containers.crun.commit", annotation->io_github_containers_crun_commit);
+  add_string_to_json (json_gen, "io.github.containers.crun.version", annotation->io_github_containers_crun_version);
+
+  yajl_gen_map_close (json_gen);
+}
+
+static inline void
+cleanup_struct_features_free (struct features_info_s **info)
+{
+  if (info == NULL || *info == NULL)
+    return;
+
+  struct features_info_s *ptr = *info;
+
+  // Free oci_version_min if it is not NULL
+  if (ptr->oci_version_min != NULL)
+    {
+      free ((void *) ptr->oci_version_min);
+      ptr->oci_version_min = NULL; // Set to NULL after freeing
+    }
+
+  // Free oci_version_max if it is not NULL
+  if (ptr->oci_version_max != NULL)
+    {
+      free ((void *) ptr->oci_version_max);
+      ptr->oci_version_max = NULL; // Set to NULL after freeing
+    }
+  if (ptr->hooks != NULL)
+    {
+      for (size_t i = 0; ptr->hooks[i] != NULL; i++)
```

Comment:
define just once `size_t i` at the beginning of the function and drop it from every for loop